src/commands.ts

```diff
@@ -323,16 +257,156 @@ export function editGroup(treeItem: vscode.TreeItem, state: State) {
       return;
     }
     const id = treeItem.id;
-    const group = state.groupArr.find(group => (group.id === id));
+    const group = state.groups.find(group => (group.id === id));
     group!.name = name;
-    refreshTreeView(state);
+    refreshFilterTreeView(state);
   });
 }
 
 export function deleteGroup(treeItem: vscode.TreeItem, state: State) {
-  const deleteIndex = state.groupArr.findIndex(group => (group.id === treeItem.id));
+  const deleteIndex = state.groups.findIndex(group => (group.id === treeItem.id));
+  if (deleteIndex !== -1) {
+    state.groups.splice(deleteIndex, 1);
+  }
+  refreshEditors(state);
+}
+
+export function saveProject(state: State) {
+  if (state.groups.length === 0) {
+    vscode.window.showErrorMessage('There is no filter groups');
+    return;
+  }
+
+  const selected = state.projects.find(p => (p.selected === true));
+  if (selected === undefined) {
+    vscode.window.showErrorMessage('There is no selected project');
+    return;
+  }
+
+  selected.groups = state.groups;
+  saveSettings(state.projects);
+
+  setStatusBarMessage(`Project(${selected.name}) is saved.`);
+}
+
+export function addProject(state: State) {
+  vscode.window.showInputBox({
+    prompt: "[PROJECT] Type a new project name",
+    ignoreFocusOut: false
+  }).then(name => {
+    if (name === undefined) {
+      return;
+    }
+
+    const project = {
+      groups: [],
+      name,
+      id: `${Math.random()}`,
+      selected: false
+    };
+
+    state.projects.push(project);
+    saveSettings(state.projects);
+    updateProjectAndTreeView(state);
+  });
+}
+
+export function editProject(treeItem: vscode.TreeItem, state: State, callback: () => void) {
+  vscode.window
+    .showInputBox({
+      prompt: "[PROJECT] Type a new name",
+      ignoreFocusOut: false,
+    })
+    .then((name) => {
+      if (name === undefined) {
+        return;
+      }
+      const findIndex = state.projects.findIndex(project => (project.id === treeItem.id));
+      if (findIndex !== -1) {
+        state.projects[findIndex].name = name;
+        saveSettings(state.projects);
+        updateProjectAndTreeView(state);
+
+        callback();
+      }
+    });
+}
+
+export function deleteProject(treeItem: vscode.TreeItem, state: State) {
+  const deleteIndex = state.projects.findIndex(project => (project.id === treeItem.id));
   if (deleteIndex !== -1) {
-    state.groupArr.splice(deleteIndex, 1);
+    if (deleteIndex === state.selectedIndex) {
+      state.groups = [];
+      state.selectedIndex = -1;
+      updateFilterAndTreeView(state);
+      refreshEditors(state);
+    }
+    state.projects.splice(deleteIndex, 1);
+    saveSettings(state.projects);
+    updateProjectAndTreeView(state);
+  }
+}
```

Comment:
It seems appropriate for the extension to handle setting the title of the view created by the extension.

`updateExplorerTitle` is reflected after the following commands:
- deleteProject
- editProject
- refreshSettings
- selectProject

Reply:
It seems that you are confused on what we are trying to convey. What we want is that instead of the current callsite of `updateExplorerTitle`, we want functions like "deleteProject" to call `updateExplorerTitle`.

Reply:
It seems more efficient for readability to handle changing the explorer view's title separately. even when the user executes `selectProject`, it would be more natural and efficient in terms of readability to call  `vscode.commands.executeCommand('workbench.view.explorer')` from extension.ts. 

The `updateExplorerTitle` function has been moved to commands.ts.